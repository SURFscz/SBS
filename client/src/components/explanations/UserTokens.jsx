import React from "react";
import I18n from "i18n-js";

export default function UserTokens() {
    if (I18n.locale === "en") {
        return [
            <section key={1} className="explanation">
                <span className="strong">Tokens</span> allow you to use the API offered by services in your own scripts.
            </section>,
            <section key={2} className="details">
                The value of the token key must be placed in
                the Authorization header of the HTTP calls performed against the API of the
                service. You create tokens per service.
            </section>,
            < section key={4} className="example">
                <code>curl -H "Authorization: bearer $token_value" "https://service/api/endpoint"</code>
            </section>
        ];
    }
        return [
            <section key={1} className="explanation">
                <span className="strong">Tokens</span> maken het mogelijk de API van diensten te gebruiken in eigen scripts.
            </section>,
            <section key={2} className="details">
                De waarde van het token moet in
                de authorization header van de HTTP-aanroep van de API van de
                dienst geplaatst worden. Gebruik een token per dienst.
            </section>,
            < section key={4} className="example">
                <code>curl -H "Authorization: bearer $token_value" "https://service/api/endpoint"</code>
            </section>
        ]

}
