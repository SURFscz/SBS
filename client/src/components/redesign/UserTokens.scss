@import "../../stylesheets/vars";
@import "../../stylesheets/forms";
@import "../../stylesheets/unit_header";

table.user-tokens {

    thead {
        th {
            &.name {
                width: 20%;
                padding-left: 15px;
            }

            &.description {
                width: 30%;
            }

            &.service {
                width: 20%;
            }

            &.expiry_date {
                width: 15%;
            }

            &.last_used_date{
                width: 15%;
            }

        }
    }

    tbody {
        td {
            &.name {
                padding-left: 15px;
            }

            &.expiry_date {
                span.expired {
                    color: var(--sds--color--red--400);
                    font-weight: bold;
                }
            }
        }
    }
}

.user-token-details-container {
    display: flex;
    max-width: $medium;
    width: 100%;
    margin: 40px auto 25px auto;

    @media (max-width: $compact-medium) {
        flex-direction: column;
        padding: 0 15px;
    }

    a.back-to-user-tokens {
        display: flex;
        align-items: flex-start;
        white-space: nowrap;
        color: $blue;
        text-decoration: underline;
        margin-right: 60px;

        svg {
            margin-right: 5px;
            fill: $blue;
            height: 22px;
        }
    }

    div.user-token-details {
        width: 100%;

        .header {
            display: flex;
            width: 100%;

            div.header-actions {
                margin-left: auto;
                display: flex;
                flex-direction: column;
                align-items: center;
            }

            button {
                margin-bottom: 15px;
            }

            span.history {
                cursor: pointer;

                svg {
                    margin-right: 10px;
                }
            }
        }

        .description {
            margin-bottom: 25px;

            &.no-header-actions {
                margin-top: 25px;
            }
        }

        .user-token-detail-actions {
            margin-bottom: 20px;
            max-width: 50%;

            div.actions__placeholder {
                color: $blue;
            }
        }
    }

    table.user-tokenMembers {
        thead {
            th {
                &.name {
                    width: 50%;
                }

                &.user__schac_home_organisation {
                    width: 35%;
                }

                &.trash, &.trash_disabled {
                    width: 74px;
                }

            }
        }

        tbody {
            td {
                &.trash {
                    text-align: right;
                    color: var(--sds--color--red--400);
                    cursor: pointer;
                }

                &.trash_disabled {
                    text-align: right;
                    color: $lighter-grey;
                    cursor: not-allowed;
                }

            }
        }

    }

    div.user-token-form {

        width: 60%;

        display: flex;
        flex-direction: column;

        @media (max-width: $compact-medium) {
            width: 100%;
            padding: 15px 15px 0 15px;
        }

        h2 {
            margin-bottom: 20px;
        }

        p.disclaimer {
            margin-bottom: 0;
        }

        div.checkbox {
            margin-top: 25px;
        }
    }

}
