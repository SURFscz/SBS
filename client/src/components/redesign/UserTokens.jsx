import React from "react";
import {
    createUserToken,
    deleteUserToken,
    reactivateUserToken,
    updateUserToken,
    userTokenGenerateValue
} from "../../api";
import {ReactComponent as ChevronLeft} from "../../icons/chevron-left.svg";
import UserTokensExplanation from "../explanations/UserTokens";
import "./UserTokens.scss";
import {isEmpty, stopEvent} from "../../utils/Utils";
import I18n from "i18n-js";
import Button from "../Button";
import {setFlash} from "../../utils/Flash";
import ConfirmationDialog from "../ConfirmationDialog";
import Entities from "./Entities";
import SpinnerField from "./SpinnerField";
import InputField from "../InputField";
import {AppStore} from "../../stores/AppStore";
import ErrorIndicator from "./ErrorIndicator";
import {dateFromEpoch, isUserTokenExpired, userTokenExpiryDate} from "../../utils/Date";
import SelectField from "../SelectField";
import DOMPurify from "dompurify";

class UserTokens extends React.Component {

    constructor(props, context) {
        super(props, context);
        const {services} = this.props;
        const serviceOptions = this.mapServicesToOptions(services)
        this.state = {
            required: ["name"],
            alreadyExists: {},
            initial: true,
            createNewUserToken: false,
            selectedUserTokenId: null,
            name: "",
            description: "",
            hashed_token: "",
            expiryDate: "",
            expired: false,
            user_id: props.user.id,
            service: serviceOptions[0],
            serviceOptions: serviceOptions,
            confirmationDialogOpen: false,
            confirmationDialogQuestion: undefined,
            confirmationDialogAction: () => true,
            cancelDialogAction: () => this.setState({confirmationDialogOpen: false}),
            loading: true,
        }
    }

    mapServicesToOptions = services => services.map(service => ({
        label: service.name,
        value: service.id,
        id: service.id,
        logo: service.logo,
        token_validity_days: service.token_validity_days
    }));


    componentDidMount = callback => {
        this.setState({
            loading: false,
            initial: true,
            createNewUserToken: false,
            selectedUserTokenId: null,
        }, callback);
    }

    refreshAndFlash = (promise, flashMsg, callback) => {
        this.setState({loading: true, confirmationDialogOpen: false});
        promise.then(res => {
            this.props.refresh(() => {
                this.componentDidMount(() => callback && callback(res));
                setFlash(flashMsg);
            });
        });
    }

    closeConfirmationDialog = () => this.setState({confirmationDialogOpen: false});

    getSelectedUserToken = () => {
        const {selectedUserTokenId} = this.state;
        const {userTokens} = this.props;
        return userTokens.find(token => token.id === selectedUserTokenId);
    }

    confirm = (action, question) => {
        this.setState({
            confirmationDialogOpen: true,
            confirmationDialogQuestion: question,
            confirmationDialogAction: action
        });
    };

    cancelSideScreen = e => {
        stopEvent(e);
        this.setState({selectedUserTokenId: null, createNewUserToken: false});
        AppStore.update(s => {
            const paths = s.breadcrumb.paths.slice(0, s.breadcrumb.paths.length - 1);
            const lastPath = paths[paths.length - 1];
            lastPath.path = null;
            s.breadcrumb.paths = paths;
        });
    }

    renderUserTokenContainer = children => {
        const {
            confirmationDialogOpen,
            cancelDialogAction,
            confirmationDialogAction,
            confirmationDialogQuestion
        } = this.state;
        return (
            <div className="user-token-details-container">
                <ConfirmationDialog isOpen={confirmationDialogOpen}
                                    cancel={cancelDialogAction}
                                    confirm={confirmationDialogAction}
                                    isWarning={true}
                                    question={confirmationDialogQuestion}/>
                <div>
                    <a className={"back-to-user-tokens"} onClick={this.cancelSideScreen} href={"/cancel"}>
                        <ChevronLeft/>{I18n.t("models.userTokens.backToUserTokens")}
                    </a>
                </div>
                {children}
            </div>
        );
    }

    newUserTokenState = (userToken, value, service) => {
        service = userToken ? this.getService(userToken) : service;
        return {
            createNewUserToken: userToken ? false : true,
            selectedUserTokenId: userToken ? userToken.id : null,
            initial: true,
            service: service,
            name: userToken ? userToken.name : "",
            description: userToken ? userToken.description : "",
            hashed_token: userToken ? userToken.hashed_token : value,
            expired: userToken ? isUserTokenExpired(userToken, service) : false,
            expiryDate: userTokenExpiryDate(userToken ? userToken.created_at : new Date().getTime() / 1000, service)
        }
    }


    renderUserTokenForm = createNewUserToken => {
        const {
            name, hashed_token, initial, description, service, serviceOptions, expiryDate, expired
        } = this.state;
        const disabledSubmit = !initial && !this.isValid();
        const children = (
            <div className="user-token-form">
                <h1>{createNewUserToken ? I18n.t("models.userTokens.new") : name}</h1>
                {createNewUserToken && <div>
                    <p className="disclaimer"
                       dangerouslySetInnerHTML={{__html: DOMPurify.sanitize(I18n.t("models.userTokens.tokenDisclaimer"))}}/>
                    <InputField value={hashed_token}
                                name={I18n.t("models.userTokens.hashedToken")}
                                toolTip={I18n.t("models.userTokens.hashedTokenTooltip")}
                                disabled={true}
                                copyClipBoard={true}/>
                </div>}
                <InputField value={name || ""}
                            onChange={e => this.setState({
                                name: e.target.value
                            })}
                            error={(!initial && isEmpty(name))}
                            placeholder={I18n.t("models.userTokens.namePlaceholder")}
                            name={I18n.t("models.userTokens.name")}
                />
                {(!initial && isEmpty(name)) && <ErrorIndicator
                    msg={I18n.t("models.userTokens.required", {
                        attribute: I18n.t("models.userTokens.name").toLowerCase()
                    })}/>}

                <InputField value={description}
                            name={I18n.t("models.userTokens.description")}
                            placeholder={I18n.t("models.userTokens.descriptionPlaceholder")}
                            onChange={e => this.setState({description: e.target.value})}
                            multiline={true}/>

                <InputField value={expiryDate}
                            disabled={true}
                            name={I18n.t("models.userTokens.expiryDate")}
                            toolTip={I18n.t("models.userTokens.expiryDateTooltip")}/>
                {expired && <ErrorIndicator
                    msg={I18n.t("models.userTokens.expiredInfo",)}/>}

                <SelectField value={service}
                             options={serviceOptions}
                             name={I18n.t("models.userTokens.service")}
                             toolTip={I18n.t("models.userTokens.serviceTooltip")}
                             onChange={this.serviceChanged}
                             searchable={false}
                />

                <section className="actions">
                    {!createNewUserToken &&
                    <Button warningButton={true}
                            onClick={this.delete}/>}
                    <Button cancelButton={true} txt={I18n.t("forms.cancel")}
                            onClick={this.cancelSideScreen}/>
                    {expired && <Button txt={I18n.t(`models.userTokens.reactivate`)}
                                        onClick={this.reactivate}/>}
                    <Button disabled={disabledSubmit} txt={I18n.t(`forms.save`)}
                            onClick={this.submit}/>
                </section>

            </div>
        );
        return this.renderUserTokenContainer(children);
    }

    serviceChanged = selectedOption => {
        const userToken = this.getSelectedUserToken();
        const newExpiryDate = userTokenExpiryDate(userToken ? userToken.created_at : new Date().getTime() / 1000, selectedOption);
        this.setState({service: selectedOption, expiryDate: newExpiryDate});
    }

    cancel = () => {
        this.setState({
            confirmationDialogOpen: true,
            leavePage: true,
            cancelDialogAction: this.gotoUserTokens,
            confirmationDialogAction: this.closeConfirmationDialog
        });
    };

    delete = () => {
        const selectedUserToken = this.getSelectedUserToken();
        const action = () => this.refreshAndFlash(deleteUserToken(selectedUserToken),
            I18n.t("userTokens.flash.deleted", {name: selectedUserToken.name}),
            () => this.setState({
                confirmationDialogOpen: false, selectedUserToken: null, editUserToken: false,
                createNewUserToken: false
            }));
        this.confirm(action, I18n.t("userTokens.deleteConfirmation", {name: selectedUserToken.name}));
    };

    isValid = () => {
        const {required, alreadyExists} = this.state;
        const inValid = Object.values(alreadyExists).some(val => val) || required.some(attr => isEmpty(this.state[attr]));
        return !inValid;
    };

    reactivate = () => {
        const {name} = this.state;
        this.refreshAndFlash(reactivateUserToken(this.stateForUserToken(false)),
            I18n.t("models.userTokens.flash.reactivated", {name: name}));

    }

    submit = () => {
        const {initial} = this.state;
        if (initial) {
            this.setState({initial: false}, this.doSubmit)
        } else {
            this.doSubmit();
        }
    };

    stateForUserToken = isNew => {
        const {name, description, hashed_token, user_id, service, selectedUserTokenId} = this.state;
        const res = {name, description, hashed_token, user_id, service_id: service.id}
        if (!isNew) {
            res.id = selectedUserTokenId
        }
        return res;
    }

    doSubmit = () => {
        if (this.isValid()) {
            const {name, createNewUserToken} = this.state;
            if (createNewUserToken) {
                this.refreshAndFlash(createUserToken(this.stateForUserToken(true)),
                    I18n.t("models.userTokens.flash.created", {name: name}));
            } else {
                this.refreshAndFlash(updateUserToken(this.stateForUserToken(false)),
                    I18n.t("models.userTokens.flash.updated", {name: name}));
            }
        } else {
            window.scrollTo(0, 0);
        }
    }

    gotoUserToken = userToken => e => {
        stopEvent(e);
        this.setState(this.newUserTokenState(userToken, userToken.hashed_token, this.getService(userToken)))
        const {collaboration, service} = this.props;
        const lastPathValue = collaboration ? `/collaborations/${collaboration.id}` : `/services/${service.id}`;
        AppStore.update(s => {
            const paths = s.breadcrumb.paths;
            const lastPath = paths[paths.length - 1];
            lastPath.path = lastPathValue;
            paths.push({
                value: I18n.t("breadcrumb.userToken", {name: userToken.name})
            })
            s.breadcrumb.paths = paths;
        });
    }

    getService = userToken => {
        const {serviceOptions} = this.state;
        return serviceOptions.find(service => service.id === userToken.service_id);
    }

    fetchNewTokenValue = () => {
        const {service} = this.state;
        userTokenGenerateValue()
            .then(res => this.setState(this.newUserTokenState(null, res.value, service)))
    }

    render() {
        const {
            loading, createNewUserToken, service
        } = this.state;
        if (loading) {
            return <SpinnerField/>;
        }
        const {userTokens} = this.props;
        const selectedUserToken = this.getSelectedUserToken();
        if (createNewUserToken || selectedUserToken) {
            return this.renderUserTokenForm(createNewUserToken);
        }

        const columns = [
            {
                key: "name",
                header: I18n.t("models.userTokens.name"),
                mapper: userToken => <a href={`${userToken.name}`}
                                        onClick={this.gotoUserToken(userToken)}>{userToken.name}</a>,
            },
            {
                key: "description",
                header: I18n.t("models.userTokens.description"),
                mapper: group => <span className={"cut-of-lines"}>{group.description}</span>
            },
            {
                key: "service__name",
                header: I18n.t("models.userTokens.service"),
                mapper: userToken => this.getService(userToken).label
            },
            {
                nonSortable: true,
                key: "expiry_date",
                header: I18n.t("models.userTokens.expiryDate"),
                mapper: userToken => {
                    const expired = isUserTokenExpired(userToken, service);
                    return <span className={`expiry_date ${expired ? "expired" : ""}`}>
                        {userTokenExpiryDate(userToken.created_at, this.getService(userToken))}
                    </span>
                }
            },
            {
                nonSortable: true,
                key: "last_used_date",
                header: I18n.t("models.userTokens.lastUsedDate"),
                mapper: userToken => {
                    return <span>
                        {userToken.last_used_date ? dateFromEpoch(userToken.last_used_date) : "-"}
                    </span>
                }
            },
        ]
        return (
            <div>
                <Entities entities={userTokens}
                          modelName="userTokens"
                          tableClassName="user-tokens"
                          searchAttributes={["name", "description"]}
                          defaultSort="name"
                          rowLinkMapper={() => this.gotoUserToken}
                          columns={columns}
                          hideTitle={true}
                          loading={loading}
                          explain={<UserTokensExplanation/>}
                          explainTitle={I18n.t("explain.tokens")}
                          showNew={true}
                          newEntityFunc={this.fetchNewTokenValue}
                          {...this.props}/>
            </div>
        )
    }
}

export default UserTokens;