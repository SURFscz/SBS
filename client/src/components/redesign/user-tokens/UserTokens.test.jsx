import React from "react";
import {render} from "@testing-library/react";
import UserTokens from "./UserTokens";

jest.mock("@fortawesome/react-fontawesome", () => ({
    FontAwesomeIcon: () => <span data-testid="icon"/>
}));

const testUserToken = {service_id: "123", created_at: 1700000000};
const testService = {id: "123", label: "Test Service", token_validity_days: 30};

describe("UserTokens", () => {
    it("renders with default props", () => {
        const {asFragment} = render(
            <UserTokens
                user={{id: "1", username: "testuser"}}
                userTokens={[testUserToken]}
                onDeleteToken={() => {
                }}
                onRevokeToken={() => {
                }}
                services={[testService]}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders with tokens", () => {
        const tokens = [
            {id: "1", name: "Token 1", created_at: 1700000000, service_id: "123"},
            {id: "2", name: "Token 2", created_at: 1700000000, service_id: "123"}
        ];
        const {asFragment} = render(
            <UserTokens
                user={{id: "1", username: "testuser"}}
                userTokens={tokens}
                onDeleteToken={() => {
                }}
                onRevokeToken={() => {
                }}
                services={[testService]}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
