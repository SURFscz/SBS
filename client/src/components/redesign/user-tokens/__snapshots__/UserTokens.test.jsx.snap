// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTokens renders with default props 1`] = `
<DocumentFragment>
  <div>
    <div
      class="mod-entities "
    >
      <section
        class="entities-search"
      >
        <h2>
          Application tokens (1)
        </h2>
        <div
          class="search-filters userTokens-search-filters"
        />
        <div
          class="search "
        >
          <div
            class="sds--text-field sds--text-field--has-icon"
          >
            <div
              class="sds--text-field--shape"
            >
              <div
                class="sds--text-field--input-and-icon"
              >
                <input
                  class="sds--text-field--input"
                  placeholder="Search for application token..."
                  type="search"
                  value=""
                />
                <span
                  class="sds--text-field--icon"
                >
                  <svg>
                    search.svg
                  </svg>
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="new-button"
        >
          <button
            class="sds--btn sds--btn--primary    "
            type="button"
          >
            <span
              class="textual"
            >
              Create application token
            </span>
          </button>
        </div>
      </section>
      <section
        class="entities-list"
      >
        <div
          class="sds--table"
        >
          <table
            class="user-tokens"
          >
            <thead>
              <tr>
                <th
                  class="name  sortable "
                >
                  Name
                  <span
                    data-testid="icon"
                  />
                </th>
                <th
                  class="description  sortable "
                >
                  Description
                </th>
                <th
                  class="service__name  sortable "
                >
                  Application
                </th>
                <th
                  class="expiry_date   "
                >
                  Expiry date
                </th>
                <th
                  class="last_used_date   "
                >
                  Last used
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="clickable "
              >
                <td
                  class="name sortable "
                >
                  <a
                    class="neutral-appearance"
                    href="undefined"
                  />
                </td>
                <td
                  class="description sortable "
                >
                  <span
                    class="cut-of-lines"
                  />
                </td>
                <td
                  class="service__name sortable "
                />
                <td
                  class="expiry_date  "
                >
                  <span
                    class="expiry_date expired"
                  >
                    December 14, 2023
                  </span>
                </td>
                <td
                  class="last_used_date  "
                >
                  <span>
                    -
                  </span>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </section>
      <div />
    </div>
  </div>
</DocumentFragment>
`;

exports[`UserTokens renders with tokens 1`] = `
<DocumentFragment>
  <div>
    <div
      class="mod-entities "
    >
      <section
        class="entities-search"
      >
        <h2>
          Application tokens (2)
        </h2>
        <div
          class="search-filters userTokens-search-filters"
        />
        <div
          class="search "
        >
          <div
            class="sds--text-field sds--text-field--has-icon"
          >
            <div
              class="sds--text-field--shape"
            >
              <div
                class="sds--text-field--input-and-icon"
              >
                <input
                  class="sds--text-field--input"
                  placeholder="Search for application token..."
                  type="search"
                  value=""
                />
                <span
                  class="sds--text-field--icon"
                >
                  <svg>
                    search.svg
                  </svg>
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="new-button"
        >
          <button
            class="sds--btn sds--btn--primary    "
            type="button"
          >
            <span
              class="textual"
            >
              Create application token
            </span>
          </button>
        </div>
      </section>
      <section
        class="entities-list"
      >
        <div
          class="sds--table"
        >
          <table
            class="user-tokens"
          >
            <thead>
              <tr>
                <th
                  class="name  sortable "
                >
                  Name
                  <span
                    data-testid="icon"
                  />
                </th>
                <th
                  class="description  sortable "
                >
                  Description
                </th>
                <th
                  class="service__name  sortable "
                >
                  Application
                </th>
                <th
                  class="expiry_date   "
                >
                  Expiry date
                </th>
                <th
                  class="last_used_date   "
                >
                  Last used
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="clickable "
              >
                <td
                  class="name sortable "
                >
                  <a
                    class="neutral-appearance"
                    href="Token 1"
                  >
                    Token 1
                  </a>
                </td>
                <td
                  class="description sortable "
                >
                  <span
                    class="cut-of-lines"
                  />
                </td>
                <td
                  class="service__name sortable "
                />
                <td
                  class="expiry_date  "
                >
                  <span
                    class="expiry_date expired"
                  >
                    December 14, 2023
                  </span>
                </td>
                <td
                  class="last_used_date  "
                >
                  <span>
                    -
                  </span>
                </td>
              </tr>
              <tr
                class="clickable "
              >
                <td
                  class="name sortable "
                >
                  <a
                    class="neutral-appearance"
                    href="Token 2"
                  >
                    Token 2
                  </a>
                </td>
                <td
                  class="description sortable "
                >
                  <span
                    class="cut-of-lines"
                  />
                </td>
                <td
                  class="service__name sortable "
                />
                <td
                  class="expiry_date  "
                >
                  <span
                    class="expiry_date expired"
                  >
                    December 14, 2023
                  </span>
                </td>
                <td
                  class="last_used_date  "
                >
                  <span>
                    -
                  </span>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </section>
      <div />
    </div>
  </div>
</DocumentFragment>
`;
